File: app/src/main/java/com/mvcoding/financius/ui/transaction/calculator/CalculatorPresenter.java
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.ui.transaction.calculator;

import android.support.annotation.NonNull;

import com.mvcoding.financius.ui.ActivityScope;
import com.mvcoding.financius.ui.Presenter;
import com.mvcoding.financius.ui.PresenterView;
import com.mvcoding.financius.util.rx.Event;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;

@ActivityScope class CalculatorPresenter extends Presenter<CalculatorPresenter.View> {
    private final Calculator calculator;

    @Inject CalculatorPresenter(@NonNull Calculator calculator) {
        this.calculator = calculator;
    }

    @Override protected void onViewAttached(@NonNull View view) {
        super.onViewAttached(view);

        final List<Observable<Event>> expressionUpdates = new ArrayList<>();
        expressionUpdates.add(view.on0Number().doOnNext(e -> calculator.digit0()));
        expressionUpdates.add(view.on1Number().doOnNext(e -> calculator.digit1()));
        expressionUpdates.add(view.on2Number().doOnNext(e -> calculator.digit2()));
        expressionUpdates.add(view.on3Number().doOnNext(e -> calculator.digit3()));
        expressionUpdates.add(view.on4Number().doOnNext(e -> calculator.digit4()));
        expressionUpdates.add(view.on5Number().doOnNext(e -> calculator.digit5()));
        expressionUpdates.add(view.on6Number().doOnNext(e -> calculator.digit6()));
        expressionUpdates.add(view.on7Number().doOnNext(e -> calculator.digit7()));
        expressionUpdates.add(view.on8Number().doOnNext(e -> calculator.digit8()));
        expressionUpdates.add(view.on9Number().doOnNext(e -> calculator.digit9()));
        expressionUpdates.add(view.onDecimal().doOnNext(e -> calculator.decimal()));
        expressionUpdates.add(view.onAdd().doOnNext(e -> calculator.add()));
        expressionUpdates.add(view.onSubtract().doOnNext(e -> calculator.subtract()));
        expressionUpdates.add(view.onMultiply().doOnNext(e -> calculator.multiply()));
        expressionUpdates.add(view.onDivide().doOnNext(e -> calculator.divide()));
        expressionUpdates.add(view.onDelete().doOnNext(e -> calculator.delete()));
        expressionUpdates.add(view.onClear().doOnNext(e -> calculator.clear()).doOnNext(e -> view.clearExpression()));
        expressionUpdates.add(view.onNumberChange().doOnNext(calculator::setNumber).map(number -> new Event()));
        expressionUpdates.add(view.onEquals()
                                      .filter(e -> !calculator.isEmptyOrSingleNumber())
                                      .map(e -> calculator.calculate())
                                      .doOnNext(calculator::setNumber)
                                      .map(number -> new Event()));

        unsubscribeOnDetach(view.onEquals()
                                    .filter(c -> calculator.isEmptyOrSingleNumber())
                                    .map(c -> calculator.calculate())
                                    .subscribe(view::startResult));

        unsubscribeOnDetach(Observable.merge(expressionUpdates)
                                    .map(e -> calculator.getExpression())
                                    .doOnNext(view::showExpression)
                                    .subscribe(expression -> {
                                        if (calculator.isEmptyOrSingleNumber()) {
                                            view.showStartResult();
                                        } else {
                                            view.showCalculate();
                                        }
                                    }));
    }

    public interface View extends PresenterView {
        @NonNull Observable<Event> on0Number();
        @NonNull Observable<Event> on1Number();
        @NonNull Observable<Event> on2Number();
        @NonNull Observable<Event> on3Number();
        @NonNull Observable<Event> on4Number();
        @NonNull Observable<Event> on5Number();
        @NonNull Observable<Event> on6Number();
        @NonNull Observable<Event> on7Number();
        @NonNull Observable<Event> on8Number();
        @NonNull Observable<Event> on9Number();
        @NonNull Observable<Event> onDecimal();
        @NonNull Observable<Event> onEquals();
        @NonNull Observable<Event> onDivide();
        @NonNull Observable<Event> onMultiply();
        @NonNull Observable<Event> onSubtract();
        @NonNull Observable<Event> onAdd();
        @NonNull Observable<Event> onDelete();
        @NonNull Observable<Event> onClear();
        @NonNull Observable<BigDecimal> onNumberChange();
        void showExpression(@NonNull String expression);
        void clearExpression();
        void showCalculate();
        void showStartResult();
        void startResult(@NonNull BigDecimal result);
    }
}

File: core-app/src/main/kotlin/com/mvcoding/expensius/feature/calculator/CalculatorPresenter.kt
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.expensius.feature.calculator

import com.mvcoding.expensius.Settings
import com.mvcoding.expensius.feature.Presenter
import com.mvcoding.expensius.feature.calculator.CalculatorPresenter.ResultDestination.TRANSACTION
import com.mvcoding.expensius.feature.calculator.CalculatorPresenter.State.CALCULATE
import com.mvcoding.expensius.feature.calculator.CalculatorPresenter.State.SAVE
import com.mvcoding.expensius.model.Transaction
import rx.Observable
import rx.Observable.merge
import rx.lang.kotlin.PublishSubject
import java.math.BigDecimal

class CalculatorPresenter(
        private val calculator: Calculator,
        private val resultDestination: ResultDestination,
        private val settings: Settings,
        private val initialNumber: BigDecimal? = null) : Presenter<CalculatorPresenter.View>() {

    private var canSave = true

    init {
        initialNumber?.let { calculator.setNumber(it) }
    }

    override fun onViewAttached(view: View) {
        super.onViewAttached(view)

        val saves = PublishSubject<Unit>()
        unsubscribeOnDetach(view.onSave().subscribe(saves))
        val expressionChanges = merge(arrayOf(
                view.onDigit0().doOnNext { calculator.digit0() },
                view.onDigit1().doOnNext { calculator.digit1() },
                view.onDigit2().doOnNext { calculator.digit2() },
                view.onDigit3().doOnNext { calculator.digit3() },
                view.onDigit4().doOnNext { calculator.digit4() },
                view.onDigit5().doOnNext { calculator.digit5() },
                view.onDigit6().doOnNext { calculator.digit6() },
                view.onDigit7().doOnNext { calculator.digit7() },
                view.onDigit8().doOnNext { calculator.digit8() },
                view.onDigit9().doOnNext { calculator.digit9() },
                view.onDecimal().doOnNext { calculator.decimal() },
                view.onAdd().doOnNext { calculator.add() },
                view.onSubtract().doOnNext { calculator.subtract() },
                view.onMultiply().doOnNext { calculator.multiply() },
                view.onDivide().doOnNext { calculator.divide() },
                view.onClear().doOnNext { calculator.clear() },
                view.onDelete().doOnNext { calculator.delete() },
                saves
                        .filter { getCurrentState() == CALCULATE }
                        .map { calculator.calculate() }
                        .doOnNext { calculator.setNumber(it); canSave = false }))

        val alteredExpressions = expressionChanges
                .startWith(Unit)
                .map { calculator.getExpression() }
                .doOnNext { view.showState(getCurrentState()) }
                .doOnNext { view.showExpression(it) }
                .map { calculator.calculate() }

        unsubscribeOnDetach(saves.filter { val result = canSave; canSave = true; result }
                .withLatestFrom(alteredExpressions, { unit, number -> number })
                .subscribe { displayResult(view, it) })
    }

    private fun getCurrentState() = if (calculator.isEmptyOrSingleNumber()) SAVE else CALCULATE
    private fun prepareTransaction(it: BigDecimal) = Transaction(currency = settings.mainCurrency, amount = it)

    private fun displayResult(view: View, amount: BigDecimal) =
            if (resultDestination == TRANSACTION) view.startTransaction(prepareTransaction(amount))
            else view.startResult(amount)

    enum class State { SAVE, CALCULATE }

    enum class ResultDestination { BACK, TRANSACTION }

    interface View : Presenter.View {
        fun onDigit0(): Observable<Unit>
        fun onDigit1(): Observable<Unit>
        fun onDigit2(): Observable<Unit>
        fun onDigit3(): Observable<Unit>
        fun onDigit4(): Observable<Unit>
        fun onDigit5(): Observable<Unit>
        fun onDigit6(): Observable<Unit>
        fun onDigit7(): Observable<Unit>
        fun onDigit8(): Observable<Unit>
        fun onDigit9(): Observable<Unit>
        fun onDecimal(): Observable<Unit>
        fun onAdd(): Observable<Unit>
        fun onSubtract(): Observable<Unit>
        fun onMultiply(): Observable<Unit>
        fun onDivide(): Observable<Unit>
        fun onDelete(): Observable<Unit>
        fun onClear(): Observable<Unit>
        fun onSave(): Observable<Unit>
        fun showExpression(expression: String)
        fun showState(state: State)
        fun startResult(number: BigDecimal)
        fun startTransaction(transaction: Transaction)
    }
}